### PR TITLE
Remove redundant API level checks, modernise some code

### PR DESCRIPTION
**Pull request by Zanthed**
@PatrykMis 
Talkback works in this PR, however because we are fully targeting API 31 now the permission `READ_PHONE_STATE` is no longer granted automatically/by default (as far as I can tell).

~~Because I don't know how to; a way to ask the user to grant access to the Phone permission (for READ_PHONE_STATE) needs to be implemented. It should be trivial.~~

Otherwise, the app will crash as the previous behaviour was it assumed it already had it so the "permission check" always returned true. READ_PHONE_STATE is considered a "dangerous" permissions and Play Store apps aren't allowed to have this by default. ~~A user-facing pop up asking to grant the Phone permissions is necessary here to prevent Talkback from crashing upon launch.~~

The check is in https://github.com/talkback-foss-team/talkback-foss/blob/main/talkback/src/main/java/com/google/android/accessibility/talkback/CallStateMonitor.java#L167

~~It needs to be made to ask the user for Phone permission and return that with error handling as it simply assumes it has it already by returning true.~~

**For accessibility purposes, the READ_PHONE_STATE permission needs to be granted by _default_.** Seems to require the `service` to be rewritten.

The crash if the permission isn't granted:
```
2021-12-05 06:17:48.850 2715-2715/com.android.talkback E/AndroidRuntime: FATAL EXCEPTION: main
    Process: com.android.talkback, PID: 2715
    java.lang.SecurityException: getCallState: Neither user 10105 nor current process has android.permission.READ_PHONE_STATE.
        at android.os.Parcel.createExceptionOrNull(Parcel.java:2425)
        at android.os.Parcel.createException(Parcel.java:2409)
        at android.os.Parcel.readException(Parcel.java:2392)
        at android.os.Parcel.readException(Parcel.java:2334)
        at com.android.internal.telecom.ITelecomService$Stub$Proxy.getCallStateUsingPackage(ITelecomService.java:2578)
        at android.telecom.TelecomManager.getCallState(TelecomManager.java:1803)
        at android.telephony.TelephonyManager.getCallState(TelephonyManager.java:5707)
        at com.google.android.accessibility.talkback.CallStateMonitor.getCurrentCallState(CallStateMonitor.java:137)
        at com.google.android.accessibility.talkback.VoiceActionMonitor.getCurrentCallState(VoiceActionMonitor.java:131)
        at com.google.android.accessibility.talkback.eventprocessor.AccessibilityEventProcessor.shouldDropEvent(AccessibilityEventProcessor.java:363)
        at com.google.android.accessibility.talkback.eventprocessor.AccessibilityEventProcessor.onAccessibilityEvent(AccessibilityEventProcessor.java:209)
        at com.google.android.accessibility.talkback.TalkBackService.onAccessibilityEvent(TalkBackService.java:533)
        at android.accessibilityservice.AccessibilityService$2.onAccessibilityEvent(AccessibilityService.java:2170)
        at android.accessibilityservice.AccessibilityService$IAccessibilityServiceClientWrapper.executeMessage(AccessibilityService.java:2376)
        at com.android.internal.os.HandlerCaller$MyHandler.handleMessage(HandlerCaller.java:44)
        at android.os.Handler.dispatchMessage(Handler.java:106)
        at android.os.Looper.loopOnce(Looper.java:201)
        at android.os.Looper.loop(Looper.java:288)
        at android.app.ActivityThread.main(ActivityThread.java:7839)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1003)
     Caused by: android.os.RemoteException: Remote stack trace:
        at android.app.ContextImpl.enforce(ContextImpl.java:2178)
        at android.app.ContextImpl.enforceCallingOrSelfPermission(ContextImpl.java:2206)
        at android.content.ContextWrapper.enforceCallingOrSelfPermission(ContextWrapper.java:905)
        at com.android.server.telecom.TelecomServiceImpl.canReadPhoneState(TelecomServiceImpl.java:2324)
        at com.android.server.telecom.TelecomServiceImpl.access$300(TelecomServiceImpl.java:78)
```

* Hard code the newest Android NDK version as it would constantly get an old version (21)
* Replaced API 30 intent flag from `PendingIntent.FLAG_UPDATE_CURRENT` to API 31's `PendingIntent.FLAG_IMMUTABLE`
* Update some dependencies
* Enable R8 full mode optimisation
* Set manifest file to app sandbox 2 and use new biometrics API
* Remove lots of redundant checks and old API hacks as we are not
  targeting API <25

Signed-off-by: June <june@eridan.me>

### CODE REVIEW

File: talkback/src/main/AndroidManifest.xml [hunk 1]
```diff
@@ -32,7 +33,8 @@
     <uses-feature android:name="android.hardware.location" android:required="false" />
 
     <!-- Required for fingerprint gesture. -->
-    <uses-permission android:name="android.permission.USE_FINGERPRINT" />
+    <uses-permission android:name="android.permission.USE_FINGERPRINT" android:maxSdkVersion="28" />
```
Comment (PatrykMis) [inline talkback/src/main/AndroidManifest.xml:36]: Why? As stated [here](https://liveaccessible.com/technology/android-accessibility/android-talkback/talkback-gestures/):

> On certain devices, notably pixel devices with finger print scanners (though some other devices may also support this), there are 4 more possible talkback gestures. Swipe up, down, left, and right on the finger print scanner. These are unassigned by default, but you can customize them to any available talkback action.

Comment (Zanthed) [reply talkback/src/main/AndroidManifest.xml:36]: Ah I didn't know USE_BIOMETRIC didn't provide fingerprint gestures.

Remove that part then.

**Comment by PatrykMis** [comment]
> because we are fully targeting API 31 now the permission READ_PHONE_STATE is no longer granted automatically/by default (as far as I can tell).

It seams for accessibility service like TalkBack, this permission is beeing granted automatically while enabling it.

**Comment by PatrykMis** [comment]
You were right. Phone permission must be granted before enabling accessibility service; otherwise, the service is constantly crashing.

**Comment by PatrykMis** [comment]
Closing as it's no longer needed, but thank you for your useful suggestions.

**Pull request closed without merging**